src/GPU3D.cpp

```diff
@@ -546,7 +550,171 @@ void GPU3D::SetEnabled(bool geometry, bool rendering) noexcept
+    FD->ZDotDisp = ((ZeroDotWLimit == 0) ? 0 : ((ZeroDotWLimit - 0x1FF) / 0x200));
+    FD->PolyAttr = CurPolygonAttr;
+    FD->PolyAttrUnset = PolygonAttr;
+    FD->VtxColor = VertexColor[2] << 10 | VertexColor[1] << 5 | VertexColor[0];
+    FD->Viewport = ((191 - Viewport[3]) & 0xFF) << 24 | Viewport[2] << 16 | ((191 - Viewport[1]) & 0xFF) << 8 | Viewport[0];
+    memcpy(FD->ProjStack, ProjMatrixStack, sizeof(ProjMatrixStack));
+    memcpy(FD->PosStack, PosMatrixStack, sizeof(PosMatrixStack));
+    memcpy(FD->VecStack, VecMatrixStack, sizeof(VecMatrixStack));
+    memcpy(FD->TexStack, TexMatrixStack, sizeof(TexMatrixStack));
+    memcpy(FD->ProjMtx, ProjMatrix, sizeof(ProjMatrix));
+    memcpy(FD->PosMtx, PosMatrix, sizeof(PosMatrix));
+    memcpy(FD->VecMtx, VecMatrix, sizeof(VecMatrix));
+    memcpy(FD->TexMtx, TexMatrix, sizeof(TexMatrix));
+    FD->MatrixMode = MatrixMode;
+    FD->Polygon = PolygonMode;
+    FD->VtxXY = CurVertex[1] << 16 | CurVertex[0];
+    FD->VtxZ = CurVertex[2];
+    FD->TexCoord = TexCoords[1] << 16 | TexCoords[0]; // use final texcoords so i dont have to worry about setting up the matrix & params properly.
+    FD->TexParam = TexParam;
+    FD->TexPalette = TexPalette;
+    FD->DiffAmbi = MatAmbient[2] << 26 | MatAmbient[1] << 21 | MatAmbient[0] << 16 | MatDiffuse[2] << 10 | MatDiffuse[1] << 5 | MatDiffuse[0];
+    FD->SpecEmis = MatEmission[2] << 26 | MatEmission[1] << 21 | MatEmission[0] << 16 | MatSpecular[2] << 10 | MatSpecular[1] << 5 | MatSpecular[0];
+    for (int i = 0; i < 32; i++)
+    {
+        u8 tbl = i << 2;
+        FD->Shininess[i] = ShininessTable[tbl];
+        FD->Shininess[i] |= ShininessTable[tbl + 1] << 8;
+        FD->Shininess[i] |= ShininessTable[tbl + 2] << 16;
+        FD->Shininess[i] |= ShininessTable[tbl + 3] << 24;
+    }
+    for (u32 i = 0; i < 4; i++)
+    {
+        FD->LightVec[i] = i << 30 | ((u32)LightDirection[i][2] & 0x3FF) << 20 | ((u32)LightDirection[i][1] & 0x3FF) << 10 | ((u32)LightDirection[i][0] & 0x3FF);
+        FD->LightColor[i] = i << 30 | LightColor[i][2] << 10 | LightColor[i][1] << 5 | LightColor[i][0];
+    }
+    FD->SwapBuffer = FlushAttributes;
+    // set writes to zero to "clear" the write list
+    FD->NumCmds = 0;
+    FD->NumParams = 0;
+    NDS.GPU.FDInProg = true;
+}
 
+void GPU3D::FinFrameDump()
+{
+    // save final latched values
+    // this step is technically completely unnecessary, does it just get optimized out?
+    FD->Disp3DCnt = RenderDispCnt;
+    for (int i = 0; i < 8; i++)
+    {
+        FD->EdgeColor[i] = RenderEdgeTable[i];
+    }
+    FD->AlphaTest = RenderAlphaRef;
+    FD->ClearColor = RenderClearAttr1;
+    FD->ClearDepOff = RenderClearAttr2;
+    FD->FogColor = RenderFogColor;
+    FD->FogOffset = RenderFogOffset;
+    for (int i = 0; i < 32; i++)
+    {
+        FD->FogTable[i] = RenderFogDensityTable[i+1];
+        FD->ToonTable[i] = RenderToonTable[i];
+    }
+
+    Platform::MakeLocalDirectory("framedumps");
```

Comment:
Instead of hardcoding a name, I'd make this method take the name as an argument. That goes for both the directory and the file itself.

Reply:
Good call, though I'll first need to figure out how to properly handle opening/browsing arbitrarily named files in the file browser of the frame dump player